frontend/ui/src/app/tutorials/branding/branding-tutorial.component.ts

```diff
@@ -0,0 +1,294 @@
+    try {
+      this.organizationBranding = await lastValueFrom(this.brandingService.get());
+    } catch (e) {
+      const msg = getFormDisplayedError(e);
+      if (msg && e instanceof HttpErrorResponse && e.status !== 404) {
+        // it's a valid use case for an organization to have no branding (hence 404 is not shown in toast)
+        this.toast.error(msg);
+      }
+    } finally {
+      this.loading.set(false);
+    }
+
+    this.brandingFormGroup.patchValue({
+      title: this.organizationBranding?.title,
+      titleDone: !!this.organizationBranding?.title,
+      description: this.organizationBranding?.description || defaultBrandingDescription,
+      descriptionDone: !!this.organizationBranding?.description,
+    });
+
+    if (!this.progress) {
+      this.loading.set(true);
+      try {
+        this.progress = await lastValueFrom(
+          this.tutorialsService.save(tutorialId, {
+            stepId: welcomeStep,
+            taskId: welcomeTaskStart,
+          })
+        );
+        this.stepper.next();
+      } catch (e) {
+        const msg = getFormDisplayedError(e);
+        if (msg) {
+          this.toast.error(msg);
+        }
+      } finally {
+        this.loading.set(false);
+      }
+    } else {
+      this.stepper.next();
+    }
+  }
+
+  protected async continueFromBranding() {
+    this.brandingFormGroup.markAllAsTouched();
+    if (this.brandingFormGroup.valid) {
+      if (this.brandingFormGroup.dirty) {
+        this.loading.set(true);
+        const formVal = this.brandingFormGroup.getRawValue();
+        const formData = new FormData();
+        formData.set('title', formVal.title);
+        formData.set('description', formVal.description);
+
+        const id = this.organizationBranding?.id;
+        let req: Observable<OrganizationBranding>;
+        if (id) {
+          req = this.brandingService.update(formData);
+        } else {
+          req = this.brandingService.create(formData);
+        }
+
+        try {
+          this.organizationBranding = await lastValueFrom(req);
+          this.brandingFormGroup.markAsPristine();
+          this.progress = await lastValueFrom(
+            this.tutorialsService.save(tutorialId, {
+              stepId: brandingStep,
+              taskId: brandingTaskSet,
+            })
+          );
+          this.toast.success('Branding options have been updated');
+        } catch (e) {
+          const msg = getFormDisplayedError(e);
+          if (msg) {
+            this.toast.error(msg);
+          }
+          return;
+        } finally {
+          this.loading.set(false);
+        }
+      }
+
+      this.brandingFormGroup.controls.titleDone.patchValue(true);
+      this.brandingFormGroup.controls.descriptionDone.patchValue(true);
+      this.prepareCustomerStep();
+      this.stepper.next();
+    }
+  }
+
+  private prepareCustomerStep() {
+    // prepare the email form
+    const email = (this.progress?.events ?? [])
+      .reverse()
```

Comment:
Note that `Array.reverse()` operates in place. To be safe, I would suggest prepending it with `.concat()` in order to ensure that we are working with a copy